edb/edgeql-parser/src/parser/mod.rs

```diff
@@ -394,6 +442,43 @@ impl<'s> Parser<'s> {
         Some(final_node.value)
     }
 
+    /// Lightweight version of act that checks if a token *could* be applied.
+    /// Returns next state.
+    fn can_act(&self, ctx: &'s Context, token: &Kind) -> Option<usize> {
+        let mut state = self.stack_top.state;
+
+        let mut node = &self.stack_top;
+        let mut pushed = 0;
+
+        loop {
+            // find next action
+            let action = ctx.spec.actions[state].get(token)?;
+
+            match action {
+                Action::Shift(next) => {
+                    return Some(*next);
+                }
+                Action::Reduce(reduce) => {
```

Comment:
We simulate Reduce in case the return is to somewhere invalid?

Reply:
No, we must simulate Reduce in case `goto` (of the reduce) takes us into a state that does not accept current token.

This is the line that will fail out if that happens: https://github.com/geldata/gel/pull/8614/files/6a2ba1437ee39283b1354ff8db0a701fb0e6a8af#diff-8edb2474c2efee3785b6c9aa287628641a7e1879b259eaaced1989d76a4db5b4R455
```
let action = ctx.spec.actions[state].get(token)?;
```